### PR TITLE
DRAFT: CDK: High level architecture

### PR DESCRIPTION
**Pull request by kmurphypolygon**
pics and commentary

### CODE REVIEW

**Review by Sharonbc01 — COMMENTED** [review]
Hi Katharine maybe we should note the following diagram depicts the full CDK Full Execution Proof (FEP) stack.

File: docs/cdk/architecture/high-level-views.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+2. The Erigon RPC node proxies the data to the CDK Erigon sequencer node.
+3. The sequencer sequences the transaction batches.
+4. The sequencer puts the batches into the data streamer component.
+5. The data streamer component streams the data into the sequence sender for validium mode stacks. It streams data into the aggregator and AggLayer for rollup and validium mode stacks. 
+6. In validium mode, the sequencer sender persists transaction data into the DAC nodes.
+7. In validium mode, the sequencer sender sequences the batches into the L1 smart contracts.
+8. In both modes, the aggregator sends the batches to the prover and the prover returns the proofs.
+9. The aggregator batches the proofs.
+10. The aggregator submits the final proof to the AggLayer.
+11. The AggLayer submits the final proof to the L1 smart contract domain.
```
Comment (Stefan-Ethernal) [inline docs/cdk/architecture/high-level-views.md]: Would it make sense to introduce some sequence diagram, that depicts this process?

File: docs/cdk/architecture/high-level-views.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+![High level view of CDK stack](../../img/cdk/cdk-stack.png)
+
+### Component interactions
+
+- External applications interact with the CDK Erigon RPC node via the transaction pool manager.
+- Engineers use a CLI tool to manage the backend components, installing and initializing various modes of operation, such as validium versus rollup for example.
+- The CDK Erigon RPC sends transactions to the CDK Erigon sequencer which sequences transactions batches.
+- The CDK Erigon sequencer puts the transaction batches into the data streamer.
+- The data streamer streams data to the sequencer sender (validium mode) and the aggregator.
```
Comment (joanestebanr) [inline docs/cdk/architecture/high-level-views.md]: I think that (validium mode) can be removed. In both modes send the same information

File: docs/cdk/architecture/high-level-views.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+- The sequencer sender sequences batches into the L1 smart contract domain and persists data into DAC nodes for validium mode operations.
+- The aggregator sends batches to the prover and receives proofs in return. It then aggregates the proofs into batches before submitting them to the L1 domain.
```
Comment (joanestebanr) [inline docs/cdk/architecture/high-level-views.md]: The aggregator send proofs to AggLayer (maybe `L1 domain` have the same meaning...)

File: docs/cdk/architecture/high-level-views.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+![High level view of CDK user data flow](../../img/cdk/user-data-flow.png)
+
+#### Sequential interactions
+
+1. User sends a transaction to the CDK Erigon RPC node.
+2. The Erigon RPC node proxies the data to the CDK Erigon sequencer node.
+3. The sequencer sequences the transaction batches.
+4. The sequencer puts the batches into the data streamer component.
+5. The data streamer component streams the data into the sequence sender for validium mode stacks. It streams data into the aggregator and AggLayer for rollup and validium mode stacks.
```
Comment (joanestebanr) [inline docs/cdk/architecture/high-level-views.md]: The same as before, doesn't mind if it's or not validium

File: docs/cdk/architecture/high-level-views.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+#### Sequential interactions
+
+1. User sends a transaction to the CDK Erigon RPC node.
+2. The Erigon RPC node proxies the data to the CDK Erigon sequencer node.
+3. The sequencer sequences the transaction batches.
+4. The sequencer puts the batches into the data streamer component.
+5. The data streamer component streams the data into the sequence sender for validium mode stacks. It streams data into the aggregator and AggLayer for rollup and validium mode stacks. 
+6. In validium mode, the sequencer sender persists transaction data into the DAC nodes.
+7. In validium mode, the sequencer sender sequences the batches into the L1 smart contracts.
```
Comment (joanestebanr) [inline docs/cdk/architecture/high-level-views.md]: That is done in both modes: validium and rollup

File: docs/cdk/architecture/high-level-views.md (hunk 3 above)
Comment (rachit77) [reply docs/cdk/architecture/high-level-views.md]:
```suggestion
- The aggregator sends batches to the prover and receives proofs in return. It then aggregates the proofs into batches before submitting them to the Agglayer or L1 depending on the chosen settlement layer.
```

How about this?

**Comment by rachit77** [comment]
[docs/img/cdk/cdk-stack.png](https://github.com/0xPolygon/polygon-docs/pull/1563/files#diff-9116eb622c2b630d7c87cc911f2ae6e7a7f2ae935b5aefb2407943480159baeb)

In this diagram, `tx-pool-manager` is connected to `sequence-sender` which is incorrect.  
`sequence-sender` should be connected to `RPC` node instead.

File: docs/img/cdk/cdk-stack.png
Comment (Stefan-Ethernal) [inline docs/img/cdk/cdk-stack.png:1]: Nitpicks:
- the font on the arrow that links the `Bridge service` and `Agglayer components` is inconsistent with the rest.
- the arrow between data-streamer and Agglayer components is oddly broken :slightly_smiling_face: Not sure if there is some auto-layout functionality, that would make the diagram look a bit cleaner.

Comment (kmurphypolygon) [reply docs/img/cdk/cdk-stack.png:1]: yeah, it's a bit hacky, thanks @Stefan-Ethernal I will endeavour to update :)

File: docs/cdk/architecture/high-level-views.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
+## CDK full execution proof (FEP)
+
+The following diagram depicts the CDK FEP stack component layout and details some of the high level communication between components.
+
+![High level view of CDK stack](../../img/cdk/cdk-stack.png)
+
+### Component interactions
+
+- Engineers use a CLI tool to manage the backend components, installing and initializing various modes of operation, such as validium versus rollup for example.
```
Comment (vcastellm) [inline docs/cdk/architecture/high-level-views.md]:
```suggestion
- Engineers use a CLI to manage the backend components, installing and initializing various modes of operation, such as validium versus rollup for example.
```

File: docs/img/cdk/cdk-stack.png
Comment (kmurphypolygon) [reply docs/img/cdk/cdk-stack.png:1]: done

File: docs/cdk/architecture/high-level-views.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+The following diagram depicts the CDK FEP stack component layout and details some of the high level communication between components.
+
+![High level view of CDK stack](../../img/cdk/cdk-stack.png)
+
+### Component interactions
+
+- Engineers use a CLI to manage the backend components, installing and initializing various modes of operation, such as validium versus rollup for example.
+- External applications send transactions to the CDK Erigon RPC node which forwards the transaction data to the sequencer via the transaction pool manager.
+- The transaction pool manager proxies manages the transactions coming in from the RPC node and streams the data onto the sequence sender.
```
Comment (vcastellm) [inline docs/cdk/architecture/high-level-views.md]: Nope, the pool manager do not communicate with the sequence-sender at all, pool manager forwards txs to the sequencer and it's the sequence-sender that reads batches from the data-stream.

Comment (vcastellm) [reply docs/cdk/architecture/high-level-views.md]: Next paragraphs are correct

File: docs/cdk/architecture/high-level-views.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+- The sequencer sequences transactions batches and puts them into the data streamer.
+- The data streamer streams data to the sequencer sender and the aggregator.
+- The sequencer sender sequences batches into the L1 smart contract domain and persists data into DAC nodes for validium mode operations.
+- The aggregator sends batches to the prover and receives proofs in return. It then aggregates the proofs into batches before submitting them to the AggLayer or L1, depending on the chosen settlement layer.
```
Comment (vcastellm) [inline docs/cdk/architecture/high-level-views.md]: "batches" can be confusing here, better just omit.

```suggestion
- The aggregator sends batches to the prover and receives proofs in return. It then aggregates the proofs before submitting them to the AggLayer or L1, depending on the chosen settlement layer.
```

File: docs/cdk/architecture/high-level-views.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+- Users interact with the bridge service via the bridge UI.
```
Comment (vcastellm) [inline docs/cdk/architecture/high-level-views.md]: or API

File: docs/cdk/architecture/high-level-views.md [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+- The AggLayer also utilizes the bridge service and interacts with the L1 smart contracts.
```
Comment (vcastellm) [inline docs/cdk/architecture/high-level-views.md]: AggLayer doesn't interact with the bridge, it only receives proofs from the cdk-node and forwards them to L1

File: docs/cdk/architecture/high-level-views.md [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+### User data flow
+
+The following diagram is a sequential depiction of the user data flow for the CDK FEP stack in validium mode using a mock prover and having an AggLayer connection.
+
+![High level view of CDK user data flow](../../img/cdk/cdk-user-data-flow.png)
+
+#### Sequential interactions
+
+1. User sends a transaction to the CDK Erigon RPC node.
```
Comment (vcastellm) [inline docs/cdk/architecture/high-level-views.md:27]: There's a gotcha here: The user can also interact directly with the pool-manager, it's not mandatory to setup an RPC node

File: docs/cdk/architecture/high-level-views.md (hunk 7 above)
Comment (kmurphypolygon) [reply docs/cdk/architecture/high-level-views.md]: DONE. Pic adjusted also.

File: docs/cdk/architecture/high-level-views.md (hunk 10 above)
Comment (kmurphypolygon) [reply docs/cdk/architecture/high-level-views.md]: I'm not sure about this @vcastellm ... https://github.com/0xPolygon/polygon-docs/blob/a30e659e665ece8a7be28673842eca335d877363/docs/agglayer/unified-bridge-api.md

File: docs/cdk/architecture/high-level-views.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+2. The Erigon RPC node proxies the data to the CDK Erigon sequencer node.
+3. The sequencer sequences the transaction batches.
+4. The sequencer puts the batches into the data streamer component.
+5. The data streamer component streams the data into the sequence sender. It streams data into the aggregator and AggLayer for rollup and validium mode stacks. 
+6. In validium mode, the sequencer sender persists transaction data into the DAC nodes.
+7. In both modes, the sequencer sender sequences the batches into the L1 smart contracts.
+8. In both modes, the aggregator sends the batches to the prover and the prover returns the proofs.
+9. The aggregator batches the proofs.
```
Comment (vcastellm) [inline docs/cdk/architecture/high-level-views.md]:
```suggestion
9. The aggregator aggregates the proofs.
```

**Comment by kmurphypolygon** [comment]
> [docs/img/cdk/cdk-stack.png](https://github.com/0xPolygon/polygon-docs/pull/1563/files#diff-9116eb622c2b630d7c87cc911f2ae6e7a7f2ae935b5aefb2407943480159baeb)
> 
> In this diagram, `tx-pool-manager` is connected to `sequence-sender` which is incorrect. `sequence-sender` should be connected to `RPC` node instead.

I'm getting conflicting reports on this @rachit77 ... perhaps we need to discuss this pic together ...

File: docs/cdk/architecture/high-level-views.md (hunk 10 above)
Comment (vcastellm) [reply docs/cdk/architecture/high-level-views.md]: WDYM? the bridge is not related to the agglayer directly

**Review by ToniRamirezM — APPROVED** [review]

File: docs/cdk/architecture/high-level-views.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
+## CDK full execution proof (FEP)
+
+The following diagram depicts the CDK FEP stack component layout and details some of the high level communication between components.
```
Comment (EmpieichO) [inline docs/cdk/architecture/high-level-views.md]:
```suggestion
The diagram below depicts a simplified architectural layout of the CDK FEP stack and indicates at a high level how components communicate.
```

File: docs/cdk/architecture/high-level-views.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+![High level view of CDK stack](../../img/cdk/cdk-stack.svg)
+
+### Component interactions
+
+- Engineers use a CLI to manage the backend components, installing and initializing various modes of operation, such as validium versus rollup for example.
```
Comment (EmpieichO) [inline docs/cdk/architecture/high-level-views.md]:
```suggestion
- The CLI tool is the starting point. Developers, or chain administrators in particular, can use the CLI tool to build and configure chains in various modes of operation, such as validiums and rollups.
```

File: docs/cdk/architecture/high-level-views.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+- External applications send transactions to the CDK Erigon RPC node which forwards the transaction data to the sequencer via the transaction pool manager.
```
Comment (EmpieichO) [inline docs/cdk/architecture/high-level-views.md]:
```suggestion
- Once a chain is configured with the CLI, users can submit transactions through the CDK Erigon RPC node. These transactions are relayed to the `tx-pool manager` before the sequencer selects and executes them.
```

File: docs/cdk/architecture/high-level-views.md [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+- The sequencer sequences transactions batches and synchronizes data with the RPC node.
```
Comment (EmpieichO) [inline docs/cdk/architecture/high-level-views.md:11]:
```suggestion
- The CDK Erigon sequencer executes the transactions, puts them in blocks and the blocks fill up batches. The CDK Erigon RPC node syncs with the CDK Erigon sequencer.
```

File: docs/cdk/architecture/high-level-views.md [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+- The sequencer sender reads batch data from the RPC node.
```
Comment (EmpieichO) [inline docs/cdk/architecture/high-level-views.md:12]:
```suggestion
- The sequencer sender reads batch data from the CDK Erigon RPC node.
```

File: docs/cdk/architecture/high-level-views.md [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+- The aggregator reads batch data from the sequencer data stream.
```
Comment (EmpieichO) [inline docs/cdk/architecture/high-level-views.md:13]:
```suggestion
- The aggregator gets batch data from the CDK Erigon sequencer via the data streamer. That is, the aggregator has a stream client with which it requests batch data from the sequencer's stream server.
```

File: docs/cdk/architecture/high-level-views.md [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+- The sequencer sender persists data into the L1 smart contract domain for rollup mode and into DAC nodes for validium mode operations.
```
Comment (EmpieichO) [inline docs/cdk/architecture/high-level-views.md:14]:
```suggestion
- Depending on whether the chain is a rollup or validium, the sequencer sender sends batch data to either smart contracts on L1 or a DAC.
```

File: docs/cdk/architecture/high-level-views.md [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+- The aggregator sends batches to the prover and receives proofs in return. Together with the prover, it aggregates the proofs into batches before submitting the final proofs to the AggLayer or L1, depending on the chosen settlement layer.
```
Comment (EmpieichO) [inline docs/cdk/architecture/high-level-views.md:15]:
```suggestion
- The aggregator sends a request to prove-batches together with the witness to the prover, and receives a proof in return. The aggregator sends proof to L1. (Note that, in the validium case, the sequence sender sends the proof together with hashes of batches to L1.)
```

File: docs/cdk/architecture/high-level-views.md [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+- Users interact with the bridge service via the bridge UI or API.
+- The AggLayer verifies proofs and interacts with the L1 smart contracts.
```
Comment (EmpieichO) [inline docs/cdk/architecture/high-level-views.md:17]:
```suggestion
- For AggLayer-connected chains, proofs from various chains get aggregated and a single final proof is sent to L1.
```

File: docs/cdk/architecture/high-level-views.md [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+
+### User data flow
+
+The following diagram is a sequential depiction of the user data flow for the CDK FEP stack in validium mode using a mock prover and having an AggLayer connection.
```
Comment (EmpieichO) [inline docs/cdk/architecture/high-level-views.md:21]: It seems this diagram is a bit redundant given the points under the "Component interactions" subsection.

File: docs/cdk/architecture/high-level-views.md [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+1. User sends a transaction to the CDK Erigon RPC node.
+2. The CDK Erigon RPC node proxies the data to the CDK Erigon sequencer node and syncs the batch data between the sequencer and the RPC nodes.
+3. The sequencer sequences the transaction batches.
+4. The sequencer sender reads batches from the RPC node.
+5. In validium mode only, the sequencer sender persists transaction data into the DAC nodes.
+6. The sequencer sender sequences the batches into the L1 smart contracts.
+7. The aggregator reads batches from the sequencer data stream.
+8. The aggregator sends batches to the provers.
+9. The aggregator submits the final proof to the AggLayer.
+10. The AggLayer submits the final proof to the L1 smart contract domain.
```
Comment (EmpieichO) [inline docs/cdk/architecture/high-level-views.md:36]: It is not clear how this flow differs from the points in the "Component interactions" subsection.

**Review by EmpieichO — APPROVED** [review]
Approving with the condition that the comments & suggestions must be addressed.

**Review by vcastellm — APPROVED** [review]

**Pull request merged**